torch/distributed/pipelining/PipelineSchedule.py

```diff
@@ -722,133 +715,172 @@ def _step_microbatches(
         total_steps = warmup_steps + fwd_bwd_steps + cooldown_steps
 
         logger.debug(
-            f"rank {self.rank}, warmup_steps {warmup_steps}, "  # noqa: G004
-            f"1f1b {fwd_bwd_steps}, cooldown_steps {cooldown_steps}"
+            "rank %s, warmup_steps %s, 1f1b %s, cooldown_steps %s",
+            rank,
+            warmup_steps,
+            fwd_bwd_steps,
+            cooldown_steps,
         )
 
-        def forward_stage_local_index(step):
-            return (step // self.pp_group_size) % self.n_local_stages
+        # Calculates the stage index based on step and pp_group_size
+        def forward_stage_index(step):
+            # Get the local index from 0 to n_local_stages-1
+            local_index = (step // self.pp_group_size) % self.n_local_stages
+            return (local_index * self.pp_group_size) + rank
 
-        def backward_stage_local_index(step):
-            return (
+        def backward_stage_index(step):
+            local_index = (
                 self.n_local_stages
                 - 1
                 - ((step - warmup_steps) // self.pp_group_size) % self.n_local_stages
             )
+            return (local_index * self.pp_group_size) + rank
+
+        fwd_stage_mb_index: Dict[int, int] = defaultdict(int)
+        bwd_stage_mb_index: Dict[int, int] = defaultdict(int)
```

Comment:
nit: document the meaning of the `key`

Reply:
nit: `defaultdict` is perhaps not a good fit here -- we don't want a `0` value (a valid mb_index) being assigned by unknown reason.